pkg/controller/clone_controller_test.go

```diff
@@ -0,0 +1,184 @@
+			cloneSource: pvcFinalizers(baseClaim(), pvcCloneFinalizer),
+			initialClaims: []runtime.Object{
+				pvcDataSourceClone(
+					pvcNamed(baseClaim(), dstName),
+					srcName,
+				),
+			},
+		},
+		"delete source pvc when destination pvc status is claim pending": {
+			cloneSource: pvcFinalizers(baseClaim(), pvcCloneFinalizer),
+			initialClaims: []runtime.Object{
+				pvcPhase(
+					pvcDataSourceClone(
+						pvcNamed(baseClaim(), dstName),
+						srcName,
+					),
+					v1.ClaimPending)},
+			expectFinalizer: true,
+		},
+		"delete source pvc when at least one destination pvc status is claim pending": {
+			cloneSource: pvcFinalizers(baseClaim(), pvcCloneFinalizer),
+			initialClaims: []runtime.Object{
+				pvcDataSourceClone(
+					pvcNamed(baseClaim(), dstName),
+					srcName,
+				),
+				pvcPhase(
+					pvcDataSourceClone(
+						pvcNamed(baseClaim(), dstName+"1"),
+						srcName,
+					),
+					v1.ClaimPending)},
+			expectFinalizer: true,
+		},
+		"delete source pvc located in another namespace should not block": {
+			cloneSource: pvcNamespaced(
+				pvcFinalizers(
+					baseClaim(), pvcCloneFinalizer,
+				),
+				srcNamespace+"1"),
+			initialClaims: []runtime.Object{
+				pvcPhase(
+					pvcDataSourceClone(
+						pvcNamed(baseClaim(), dstName+"1"),
+						srcName,
+					),
+					v1.ClaimPending)},
+		},
+		"delete source pvc which is not cloned by any other pvc": {
+			cloneSource: pvcFinalizers(baseClaim(), pvcCloneFinalizer),
+		},
+		"delete source pvc without finalizer": {
+			cloneSource: baseClaim(),
+		},
+	}
+
+	for k, tc := range testcases {
+		tc := tc
+		t.Run(k, func(t *testing.T) {
+			t.Parallel()
+			var clientSet *fakeclientset.Clientset
+
+			utilruntime.ReallyCrash = false
+
+			objects := append(tc.initialClaims, tc.cloneSource)
+			clientSet = fakeclientset.NewSimpleClientset(objects...)
+			informerFactory := informers.NewSharedInformerFactory(clientSet, 1*time.Second)
+			claimInformer := informerFactory.Core().V1().PersistentVolumeClaims().Informer()
+			claimLister := informerFactory.Core().V1().PersistentVolumeClaims().Lister()
+			rateLimiter := workqueue.NewItemExponentialFailureRateLimiter(time.Second, 2*time.Second)
+			claimQueue := workqueue.NewNamedRateLimitingQueue(rateLimiter, "claims")
+
+			for _, claim := range tc.initialClaims {
+				claimInformer.GetStore().Add(claim)
+			}
+
+			informerFactory.WaitForCacheSync(context.TODO().Done())
+			go informerFactory.Start(context.TODO().Done())
+
+			cloningProtector := NewCloningProtectionController(
+				clientSet,
+				claimLister,
+				claimInformer,
+				claimQueue,
+			)
+
+			go cloningProtector.Run(1, context.TODO().Done())
+
+			// Simulate Delete behavior
+			claim := tc.cloneSource.(*v1.PersistentVolumeClaim)
+			clientSet.CoreV1().PersistentVolumeClaims(claim.Namespace).Update(pvcDeletionMarked(claim))
+
+			// Wait for couple reconciles for controller to adjust finalizers
+			time.Sleep(2 * time.Second)
```

Comment:
Is it necessary to wait for a couple of reconciles? For an unit test we typically call `syncClaim()` and don't run whole controller loop with workqueues.

(Yes, there will be some functions not covered by unit tests.)

Reply:
No, not at all. Will fix this, thanks for suggestion. It will also allow to catch an error, which forces the reconcile to run again.